### PR TITLE
Implement new retry strategy

### PR DESCRIPTION
**Pull request by maxiloc**

### CODE REVIEW

**Comment by JanPetr** [comment]
LGTM

File: src/AlgoliaSearch/Client.php [hunk 1]
```diff
@@ -704,31 +704,32 @@ public static function buildQuery($args)
...
+     * @internal param int $connectTimeout
      */
     public function request(
         $context,
         $method,
         $path,
         $params,
         $data,
-        $hostsArray,
+        HostsHandler $hostsArray,
```
Comment (rayrutjes) [inline src/AlgoliaSearch/Client.php]: This introduces a backward compatibility issue.

Comment (maxiloc) [inline src/AlgoliaSearch/Client.php]: Yes we'll need to do a 2.0

Comment (maxiloc) [inline src/AlgoliaSearch/Client.php]: Ok I'll find a way

**Comment by maxiloc** [comment]
@redox @rayrutjes @JanPetr Can you have a look again, I made it BC

**Comment by redox** [comment]
> @redox @rayrutjes @JanPetr Can you have a look again, I made it BC

The only issue I see is that using such `static` attribute is not compatible with having several API clients with different `application_id` and/or `hosts`; right?

**Comment by maxiloc** [comment]
We need it static to be able to handle the strategy when you instanciate twice with the same app id and host. But we can make it in a way that handle different app id and host

**Comment by rayrutjes** [comment]
I have a solution underway. Should be there in a few.

**Pull request closed without merging**